### DISCUSSION TITLE
Discussion about generated C code

### DISCUSSION THREAD

**Discussion by stefanos82** [edubart/nelua-lang#53] [General]
I run the `examples/helloworld.nelua` and I have noticed in generated C code the following header files:

![image](https://user-images.githubusercontent.com/1777466/106531935-d5b46080-64f7-11eb-8ced-80671f3119d7.png)

I have looked for the use of `bool` anywhere in the generated code and could not find it in there and wondered why it was even added there :thinking: 

Also, if we open `<inttypes.h>` header file, we will see that it calls `<stdint.h>` as well; so, we don't need to add it during generating code.

![image](https://user-images.githubusercontent.com/1777466/106532065-2926ae80-64f8-11eb-810a-85e7787c1b5f.png)

The same goes with `<stddef.h>` which `<stdio.h>` includes at the very first lines of it.

I know `ifdef` guards they take place and ignore the extra calls, but would it hurt if we clean the generated code?

...unless it's something that is only part of GCC, the compiler of my personal choice.

Now that's interesting; let me check by using Clang instead of GCC.

**Comment by stefanos82** [comment]
Yeah, my original suspicion was correct.

It's a GCC "behavior", so to speak; clang does not do such thing with its header files.

**Comment by stefanos82** [comment]
I have tested both GCC and Clang with the following `#if` condition and work as expected.

![image](https://user-images.githubusercontent.com/1777466/106535086-18793700-64fe-11eb-9094-0041774764ab.png)

Yeah, I deleted `<stdbool.h>` to check whether some magic with boolean was taking place but... :shrug: nothing!

```sh
clang -o helloworld helloworld.c -Wall -fwrapv -g -lm -s
gcc -o helloworld helloworld.c -Wall -fwrapv -g -lm -s
```

**Comment by edubart** [comment]
> The same goes with <stddef.h> which <stdio.h> includes at the very first lines of it.

The C generate code follow rules from the C standard. For example, the standard say if you need to use `NULL` you must include `stddef.h`, it says nothing that this can be omitted in case some other file is already included, if GCC does this it's an implementation detail of GCC, but Nelua has to work with other C compilers like Clang, GCC, TCC, so it cannot generate C code dependent on implementation details.

> I have looked for the use of bool anywhere in the generated code and could not find it in there and wondered why it was even added there.

Because the following headers are exceptions, and are always included in any generated C code, as they are required for Nelua primitives types: 
* stddef.h
* stdint.h
* stdbool.h

But yea, the Nelua compiler could include them only when really used as a small enhancement, but in general, in usual applications, they will be always included anyway.

**Comment by edubart** [comment]
I've improved the C code generation to include those headers only when required in https://github.com/edubart/nelua-lang/commit/c0fdc425dc42a33ab06d4e18158ad0b0d28797f9 and  https://github.com/edubart/nelua-lang/commit/33d36d93dcc5319dccd29c719263b834dcaf9ad9

In summary now:

* `stddef.h` is included when using `NULL` , `size_t` or `ptrdiff_t`, as described in https://www.cplusplus.com/reference/cstddef/ .
* `stdint.h` is included when using fixed sized integers (`intX_t` or `uintX_t`), as described in https://www.cplusplus.com/reference/cstdint/.
* `stdbool.h` is included when using the `bool` type, as described in https://www.cplusplus.com/reference/cstdbool/.
* `inttypes.h` is included when using the `PRIx` macros, as described in https://www.cplusplus.com/reference/cinttypes/.

**Comment by stefanos82** [comment]
I have just tested it; it's more than awesome, kudos!

![image](https://user-images.githubusercontent.com/1777466/106621974-c676f680-657b-11eb-92de-ef5c28e79394.png)

So...when you are going to write "The Design and Implementation of Nelua"?

I must have it in my collection! :laughing:

**Comment by stefanos82** [comment]
About the use of `NULL`; the standard says "...expands to an implementation-defined null pointer constant".

We do know though that some compilers use a plain `0` instead of `(void *)0`.

In my humble opinion, I would not want to include an entire header file for using a single thing, such as `NULL` which could be easily defined inside my generated code.

Personally, what I would do would be to undefined `NULL` and redefine it as `#define NULL ((void *)0)` to meet standard's credential **and** be supported on all platforms out there.

Thoughts?

**Reply by edubart** [reply]
A goal of Nelua is to generate readable and clear C code, and a C developers expects NULL to be defined by `stddef.h`. I don't see the point of avoiding including these headers, it will be included anyway to have the `size_t` type which is usually used.

Also I want to keep the generated C file simple and small as possible, so I try to not add new defines. For example the section disabling warnings and defining `nelua_static_assert` is already bothers me a little, but they are required for a good user experience.

**Reply by stefanos82** [reply]
> A goal of Nelua is to generate readable and clear C code, and a C developers expects NULL to be defined by `stddef.h`. I don't see the point of avoiding including these headers, it will be included anyway to have the `size_t` type which is usually used.
> 

Ah yes, I forgot about the use of `size_t`; my bad.

> Also I want to keep the generated C file simple and small as possible, so I try to not add new defines. For example the section disabling warnings and defining `nelua_static_assert` is already bothers me a little, but they are required for a good user experience.

Interesting.

So, what if I did not want to have a readable C code, but wanted to have a highly performant binary output, what direction should I take, the hand-tweaking or the flag-usage?

That is, assuming there's a flag available that fully optimizes the C code to push its limits towards the best performance possible.

My personal preference would be: write everything you want in Nelua and tweak the generated C code when you must; and that should be only the exception of 1% to 5% of my whole time.

If Nelua comes with a set of flags and methodologies to do the dirty work for me behind the scenes, even better!

**Reply by edubart** [reply]
> So, what if I did not want to have a readable C code, but wanted to have a highly performant binary output, what direction should I take, the hand-tweaking or the flag-usage?

First note that these includes does not impact the application performance, and Nelua already prioritizes performance in the C generated code over readability. 

For maximum performance you should:

* Compile in release mode, that is `--release` flag, this will pass `-O2 -fno-plt` to the C compiler.
* Disable runtime checks, that is`--pragma nochecks` flag, this will not emit bounds checking code, null pointer dereferencing checks, or checks for numeric overflow when casting, plus other checks. It's better to have them enabled when developing though to catch mistakes.
* Disable the GC only if you are doing manual memory management, that is `--pragma nogc` flag, but note this is very dependent on your application and coding style.
* Optionally pass -O3 to the C compiler, that is `--cflags="-O3"` flag, but note this is application dependent, some applications are faster using just -O2.
* There are other interesting C flags like `-ffast-math` `-flto` and `-march=native`, but I will not speak about them here.

In summary you could get **maximum performance** doing something like this with many **remarks**:

```bash
nelua --release --pragma nochecks --pragma nogc --cflags="-O3 -ffast-math -flto -march=native" helloworld.nelua
```

> My personal preference would be: write everything you want in Nelua and tweak the generated C code when you must; and that should be only the exception of 1% to 5% of my whole time.

In Nelua you can emit raw C code anytime in case you need a tweak, but usually you don't need that because Nelua have most C features without adding overhead.

**Reply by stefanos82** [reply]
Good to know all these things.

I went through the entire docs and couldn't find any thorough explanation of Nelua's flags.

Are there any plans to add them in the documentation?

I guess for now I could go with the traditional "trial and error" way, by using the `-h, --help` flag and experiment accordingly.

Is this the best approach for now or is there any other way you would suggest to go otherwise?

**Reply by edubart** [reply]
Note that the language is still under development, thus you will find some stuff incomplete like the documentation, the flags are not documented and at the moment you have to read the `--help` output, the standard libraries are not documented much too and you have to read their source files in `lib` folder. If you have daily questions due to missing documentation me and others can answer in discord's in the #help channel.

**Reply by stefanos82** [reply]
Do I need an account to join the channel? I don't use Discord myself, but I do use IRC.
